shared-module/adafruit_bus_device/I2CDevice.c

```diff
@@ -31,48 +31,75 @@
 #include "py/runtime.h"
 #include "lib/utils/interrupt_char.h"
 
-void common_hal_adafruit_bus_device_i2cdevice_construct(adafruit_bus_device_i2cdevice_obj_t *self, busio_i2c_obj_t *i2c, uint8_t device_address) {
+void common_hal_adafruit_bus_device_i2cdevice_construct(adafruit_bus_device_i2cdevice_obj_t *self, mp_obj_t *i2c, uint8_t device_address) {
     self->i2c = i2c;
     self->device_address = device_address;
 }
 
 void common_hal_adafruit_bus_device_i2cdevice_lock(adafruit_bus_device_i2cdevice_obj_t *self) {
-    bool success = common_hal_busio_i2c_try_lock(self->i2c);
+    mp_obj_t dest[2];
+    mp_load_method(self->i2c, MP_QSTR_try_lock, dest);
 
-    while (!success) {
+    mp_obj_t success = mp_call_method_n_kw(0, 0, dest);
+
+    while (!mp_obj_is_true(success)) {
         RUN_BACKGROUND_TASKS;
         if (mp_hal_is_interrupted()) {
             break;
         }
 
-        success = common_hal_busio_i2c_try_lock(self->i2c);
+        success = mp_call_method_n_kw(0, 0, dest);
     }
 }
 
 void common_hal_adafruit_bus_device_i2cdevice_unlock(adafruit_bus_device_i2cdevice_obj_t *self) {
-    common_hal_busio_i2c_unlock(self->i2c);
-}
-
-uint8_t common_hal_adafruit_bus_device_i2cdevice_readinto(adafruit_bus_device_i2cdevice_obj_t *self, mp_obj_t buffer, size_t length) {
-    return common_hal_busio_i2c_read(self->i2c, self->device_address, buffer, length);
-}
-
-uint8_t common_hal_adafruit_bus_device_i2cdevice_write(adafruit_bus_device_i2cdevice_obj_t *self, mp_obj_t buffer, size_t length, bool transmit_stop_bit) {
-    return common_hal_busio_i2c_write(self->i2c, self->device_address, buffer, length, transmit_stop_bit);
+    mp_obj_t dest[2];
+    mp_load_method(self->i2c, MP_QSTR_unlock, dest);
+    mp_call_method_n_kw(0, 0, dest);
 }
 
 void common_hal_adafruit_bus_device_i2cdevice_probe_for_device(adafruit_bus_device_i2cdevice_obj_t *self) {
     common_hal_adafruit_bus_device_i2cdevice_lock(self);
 
-    mp_buffer_info_t bufinfo;
-    mp_obj_t buffer = mp_obj_new_bytearray_of_zeros(1);
+    mp_buffer_info_t write_bufinfo;
+    mp_obj_t write_buffer = mp_obj_new_bytearray_of_zeros(0);
+    mp_get_buffer_raise(write_buffer, &write_bufinfo, MP_BUFFER_READ);
+
+    mp_obj_t dest[4];
 
-    mp_get_buffer_raise(buffer, &bufinfo, MP_BUFFER_WRITE);
+    /* catch exceptions that may be thrown while probing for the device */
+    nlr_buf_t write_nlr;
+    if (nlr_push(&write_nlr) == 0) {
+        mp_load_method(self->i2c, MP_QSTR_writeto, dest);
+        dest[2] = mp_obj_new_int_from_ull(self->device_address);
+        dest[3] = write_buffer;
+        mp_call_method_n_kw(2, 0, dest);
+        nlr_pop();
+    } else {
+        /* some OS's don't like writing an empty bytestring... retry by reading a byte */
```

Comment:
"some OS's" -- this may be for blinka support.  See https://github.com/adafruit/Adafruit_CircuitPython_BusDevice/pull/22 -- I don't think any I2C driver in CircuitPython has this limitation, because the internal probe function (used only for scan, and not available to call here, sadly) doesn't have the second detection method.
> It was writing a zero byte value to the bus. This triggered a bug in the Linux kernel I2C bus OMAP2 driver for the BeagleBone. The driver returns an invalid write error when msg->len is 0: [...]

Reply:
That makes sense to me. I can remove the section that uses read after the write fails if it will not be used in the core. Does that make sense?

Reply:
I think so.  You get to test it :)  thanks.